enhancements/api-review/add-ImagePolicy-to-config.openshift.io-v1alpha1.md

```diff
@@ -0,0 +1,294 @@
+	Images []ImageScope `json:"images"`
+	// policy defines the verification policy
+	Policy Policy `json:"policy"`
+}
+
+// ImageScope is the item of the images list.
+// https://go.dev/play/p/FfL_3bpqSX5
+// +kubebuilder:validation:Pattern=`^\*(?:\.(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]))+$|^((?:(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])(?:(?:\.(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]))+)?(?::[0-9]+)?/)?[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?(?:(?:/[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?)+)?)(?::([\w][\w.-]{0,127}))?(?:@([A-Za-z][A-Za-z0-9]*(?:[-_+.][A-Za-z][A-Za-z0-9]*)*[:][[:xdigit:]]{32,}))?$`
+type ImageScope string
+
+type Policy struct {
+	// oidcIssuer contains the expected OIDC issuer.
+	// Example: "https://expected.OIDC.issuer/"
+	// Required if fulcioCAData is set.
+	// subjectEmail must be set and keyData must be empty if oidcIssuer is set
+	// +optional
+	OIDCIssuer string `json:"oidcIssuer,omitempty"`
+
+	// subjectEmail holds the email address of the subject.
+	// Example: "expected-signing-user@example.com"
+	// Required if fulcioCAData is set.
+	// oidcIssuer must be set and keyData must be empty if subjectEmail is set.
+	// +optional
+	SubjectEmail string `json:"subjectEmail,omitempty"`
+
+	// keyData contains inline base64 data of the public key.
+	// keyData can be empty if the image got signed keyless.
+	// Requires oidcIssuer, subjectEmail, and fulcioCAData to be empty if keyData is set.
+	// +optional
+	KeyData string `json:"keyData,omitempty"`
+
+	// fulcioCAData contains inline base64 data for the fulcio CA certificate.
+	// +optional
+	FulcioCAData string `json:"fulcioCAData,omitempty"`
+
+	// rekorKeyData contains inline base64 data of the rekor public key.
+	// Required if fulcioCAData is set.
+	// +optional
+	RekorKeyData string `json:"rekorKeyData,omitempty"`
+
+	// signedIdentity specifies what image identity the signature.
+	// claims about the image.
+	// +optional
+	SignedIdentity Identity `json:"signedIdentity,omitempty"`
+}
+
+type Identity struct {
+	// Defaults to "MatchRepoDigestOrExact"
+	IdentityMatchPolicy IdentityMatchPolicy `json:"identityMatchPolicy,…"`
+	DockerRepository    string              `json:"dockerRepository,omitempty"`
+	Prefix string `json:"prefix,omitempty"`
+	SignedPrefix string `json:"signedPrefix,omitempty"`
+}
+
+type IdentityMatchPolicy string
+
+const (
+	IdentityMatchPolicyMatchRepoDigestOrExact = "MatchRepoDigestOrExact"
+	IdentityMatchPolicyMatchRepository        = "MatchRepository"
+	IdentityMatchPolicyExactRepository        = "ExactRepository"
+	IdentityMatchPolicyRemapIdentity          = "RemapIdentity"
+)
+
+type ImagePolicyStatus struct {
+	// PolicyJSON contains the whole policy applied to the namespace
+	// which got written to disk. This includes cluster-wide policies
+	// from the `openshift-config` namespace as well.
+	PolicyJSON string `json:"policyJSON,omitempty"`
```

Comment:
Merged policy will be written to all of `ImagePolicy` objects. MCO reads all the `ImagePolicy` objects and writes the policies to a machine config.

Reply:
What happens if the CRs applying to a single namespace contain not-exactly-identical `ImagePolicy.Status.PolicyJSON` values? At least in the naive implementation, and IIRC in the current MCO code organization, the controller that turns `ImagePolicySpec` into `ImagePolicyStatus`, and the controller than turns `ImagePolicyStatus` into `MachineConfig`, run concurrently without coordination, so it’s not obvious to me that this can be ruled out.

Hopefully that should be transient state, but still… if the `PolicyJSON` field can not be trusted and admins debugging need to investigate `MachineConfig` directly, is there a value in all those data copies?

Reply:
I share the concerns about copying this data many times. The MCO should have a single source of truth for this

Reply:
I think that the PolicyJSON field will be in its intended state once the CRs are settled. 
 we can also consider the possibility of removing this field and guiding administrators to check `<namespace>.json`, which contains both the cluster-level override and namespace-specific policies within the `99-<pool>-imagepolicy machine-config`.
@saschagrunert wdyt?

Reply:
@yuqi-zhang 
Should we guide users to access the machine configuration in order to retrieve the configuration? Based on the current design, I don't know whether there is another alternative location where the consolidated policy can be centrally stored.

Reply:
@QiWang19 sounds good to me! We can add the field later on if required.

Reply:
> Should we guide users to access the machine configuration in order to retrieve the configuration? Based on the current design, I don't know whether there is another alternative location where the consolidated policy can be centrally stored.

As I understand based on the discussion, what is being proposed is either:
1. the MCC merges all policies into the 99-pool-imagepolicy config, which acts as the only source of truth for the merged policy, or
2. something merging policies first into a (configmap?) and then the MCO just renders based on the configmap and doesn't actually understand policy merging logic (but who else would do this)?

I think either one should work in theory, I don't have a strong feeling, assuming both implementations use the MCO as it's rendering path